src/collected_output.rs

```diff
@@ -55,12 +65,16 @@ impl Waiter {
             Ok(collected_stderr)
         });
         Waiter {
+            stdin: stdin_join_handle,
             stdout: stdout_join_handle,
             stderr: stderr_join_handle,
         }
     }
 
     pub(crate) fn join(self) -> io::Result<CollectedOutput> {
+        self.stdin
+            .join()
+            .expect("stdout relaying thread panicked")?;
```

Comment:
Can this return an `io::Error` instead of panicking?

If the child process doesn't read from stdin, would this cause the `.expect` to panic?

Reply:
> Can this return an `io::Error` instead of panicking?

No. Well, it's a bit complicated. There's two layers of `Result`s here:
The call to `join` (in this context here) returns `std::thread::Result<io::Result<()>>`.
The outer `thread::Result` layer is introduced by [`JoinHandle::join`](https://doc.rust-lang.org/std/thread/struct.JoinHandle.html#method.join) (the one that's being called here).
That `join` returns an `Err` if and only if the child thread panics.
The error type in there is `Box<dyn Any + Send + 'static>` and I don't think that can be converted to an `io::Error`.
Also, in the case where the thread panics, I think it's best to re-throw that panic in the main thread.
Instead of just letting the main thread continue, even though something went terribly wrong.

The inner `Result` layer here is the `io::Result` that is being produced by the closure passed to `std::thread::spawn`.
And that `Result` is *not* converted into a panic, but errors are just propagated normally. (As in `Waiter::join` returns an `io::Result`.)

Hope this explanation helps and that's what you were asking.

> If the child process doesn't read from stdin, would this cause the `.expect` to panic?

No. If the child process doesn't read from stdin, and the parent writes to it, then it'll probably go into the OS buffer (I think) and the writing thread will exit normally. If that buffer is full, one of the `write` calls in the writing thread will block. And then the `join` call would hang. :scream: I'll have to test that. Good catch...

Reply:
I added return types to the closures for the three threads in here in hope to make it a bit clearer what's going on with the error handling.

Also, I pushed [the `stdin-experiment` branch](https://github.com/soenkehahn/cradle/compare/063135974db11e0f69f132595f64f4dc0b94d507...stdin-experiment) to test the behavior when
- the parent writes a lot of bytes into the child's `stdin`,
- the child terminates without reading from `stdin`.

I suspected that the `write` call would hang indefinitely in the writing thread, causing this `JoinHandle::join` call to hang. But that's not what happens. Instead when the child terminates, the read end of the pipe gets closed and the blocking write call errors with an `io::Error`. (That error will then get propagated properly, meaning `Waiter::join` returns an `Err`. So i.e. using `cmd_result!` will allow users to handle that error through `Result`.)

I think that that might be exactly the behavior we want. If users try to write a bunch of stuff into the child's `stdin`, but that process never reads from it and instead closes `stdin` (for example by terminating), it seems fair to throw an error. It's a bit weird that there's the OS buffer in between, which means that if the parent writes less bytes to the child's `stdin`, the error never shows up. But there's little we can do about that. I think this is just how pipes work. Or is there a different, more reasonable behavior?

In any case we should add a test for documenting whatever the behavior is.

(I also did a similar experiment where the parent fills the buffer of the child's `stdin`, the child never reads from it, but it also never terminates. In that case the writing thread just hangs on one of the `write` calls. And the main thread hangs on waiting for the child process to terminate. Which seems like correct behavior.)

Reply:
I now added this test: https://github.com/soenkehahn/cradle/pull/59/commits/bff35517d01862e680b4f2b203b1cbb51eb8aad2

Reply:
Gotcha, that all makes sense. Correct me if I'm wrong, but I think we now don't think that any of the worker threads can panic?

If so, then it's reasonable to leave the `.expect` in, since we think it can't be triggered.

Another option, not sure how valuable it is, would be to map the err value of `std::thread::Result` to an `io::Error`, using `std::io::Error::new(std::io::ErrorKind::Other, join_error)`.

This would allow propagating the join error, if it were to occur, as an io::Error.

Reply:
> It's a bit weird that there's the OS buffer in between, which means that if the parent writes less bytes to the child's stdin, the error never shows up. But there's little we can do about that. I think this is just how pipes work. Or is there a different, more reasonable behavior?

I missed this question. This seems fine to me. The user requested that we write some bytes to the child's stdin, but the child didn't read them all, that seems okay, since there are lots of circumstances in which a command might not read all of its input.

Reply:
> Gotcha, that all makes sense. Correct me if I'm wrong, but I think we now don't think that any of the worker threads can panic?

Yes, exactly.